Unity-Technologies/ScriptableRenderPipeline
# Rework Sub Graph importer

### Purpose of this PR
Remove SubGraphDatabase in favour of using `GatherDependenciesFromSourceFile` in importers for handling nested sub graphs. Currently, the sub graph importer still has to detect mutually recursive sub graphs, because of how they are handled by the import pipeline.

To make this faster, I've introduced `MinimalGraphData` which only deserialises exactly what it needs to determine dependencies.

`GetSourceAssetDependencies(List<string>)` has been removed from `AbstractMaterialNode`. Instead, nodes can provide a minimal type via `HasDependenciesAttribute`.

I tried to make Git track the renames nicely, but GitHub isn't following. Essentially I've merged `SubGraphAsset` and `SubGraphData`, as well as `ShaderSubGraphImporter` and `SubGraphDatabaseImporter`.

Fixes [FB1138262: Recursive Sub Graph check & error badge doesn't happen until file reload](https://fogbugz.unity3d.com/f/cases/1138262/). This issue turned out to be worse than assumed. What actually happened was that existing Shader Graph windows would get stuck on old sub graph data. We now reload sub graph data in all open Shader Graph windows after any sub graph has been imported.

Fixes [FB1076181: Having multiple tabs open and in "focus" causes multiple (maybe all) open assets to compile](https://fogbugz.unity3d.com/f/cases/1076181/). There is a separate issue with ADBv2 that will cause all Shader Graphs using Sub Graphs to re-compile their preview shader when any Sub Graph is saved. For ADBv1 users this is not an issue.

---
### Testing status
**Katana Tests**: Green except for unrelated VFX failure https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?unity_branch=trunk&ScriptableRenderLoop_branch=sg/fix-1138262&automation-tools_branch=master

**Manual Tests**: What did you do?
- Ensure no repro without fix
- Ensure repro with fix

**Automated Tests**: nil

---
### Overall Product Risks
**Technical Risk**: Medium

**Halo Effect**: Low
